### PR TITLE
Geomap: Support shared crosshair for route layer

### PR DESCRIPTION
**Pull request by alexanderzobnin**
This PR adds support for the shared crosshair feature for the route visualization layer. In order to use this feature, data frame with geo data should contain time field. Any feedback appreciated. Ref #43554

https://user-images.githubusercontent.com/4932851/176259999-f05873ef-a93b-40e3-8e25-eee5a0f6af9b.mov

### CODE REVIEW

**Review by nmarrs — COMMENTED** [review]
This looks great! I'm having a little trouble testing it out locally though. My current issue is that when I pass basic csv data (lat, lng, time) the point object only persists the lat and lng properties causing `time` to be undefined during DataHoverEvent callback

Any thoughts on this / would it be possible to have a sample of the data you are using to get this to work? Thanks!

File: public/app/plugins/panel/geomap/layers/data/routeLayer.tsx [hunk 1]
```diff
@@ -94,8 +101,69 @@ export const routeLayer: MapLayerRegistryItem<RouteConfig> = {
...
+    // Crosshair layer
+    const crosshairFeature = new Feature({});
+    const crosshairStyle = new Style({
+      image: new Circle({
+        radius: 6,
+        stroke: new Stroke({
+          color: 'rgb(253, 253, 253)',
+          width: 1.5
+        }),
+        fill: new Fill({color: 'rgb(242, 204, 12)'}),
```
[the remark this answers is not in the document]
Comment (alexanderzobnin) [reply public/app/plugins/panel/geomap/layers/data/routeLayer.tsx:113]: Yes, good point.

**Comment by alexanderzobnin** [comment]
@nmarrs I use it with Strava data source. So I added time field to the geo data to make it work. Otherwise it's not possible to link geo points to the time line.

**Comment by nmarrs** [comment]
Thanks for providing the test data! Played around a bit with the provided test data but am still having a hard time getting it work as expected locally. It seems that maybe the DataHoverEvent is only firing when hovering over the geomap panel vs the time series panel as the event's point does not have a `time` property. Geomap csv data has time field that matches time on time series

https://user-images.githubusercontent.com/22381771/176335002-ba179dda-a2ed-448a-9a29-35e3c0ec24dd.mov

**Comment by nmarrs** [comment]
Something I noticed while testing with crosshair data, the route opacity editor is inverted (defaulting at 0 instead of expected 1)

https://user-images.githubusercontent.com/22381771/176335482-cf052b63-714e-479a-855a-e48dc4202842.mov

**Comment by alexanderzobnin** [comment]
@nmarrs Opacity sliders really work strange, but it's not under layer control. The first one - is color opacity config, which is from default <StyleEditor>. The second one is a kind of "Layer opacity" which is not controlled by the routeLayer - it works in opposite way for some reason and works in the same way for another layers (Heatmap, GeoJSON, etc).

File: public/app/plugins/panel/geomap/GeomapPanel.tsx [hunk 2]
```diff
@@ -590,11 +590,7 @@ export class GeomapPanel extends Component<Props, State> {
         if (v.id === MapCenterID.Coordinates) {
           coord = [config.lon ?? 0, config.lat ?? 0];
         } else if (v.id === MapCenterID.Fit) {
-          var extent = layers
-            .getArray()
-            .filter((l) => l instanceof VectorLayer)
-            .map((l) => (l as VectorLayer<any>).getSource().getExtent() ?? [])
-            .reduce(extend, createEmpty());
+          const extent = getLayersExtent(layers);
```
Comment (nmarrs) [inline public/app/plugins/panel/geomap/GeomapPanel.tsx:593]: much cleaner 🧼 :)

**Review by nmarrs — APPROVED** [review]
Works well! Great job! 💯

**Comment by nmarrs** [comment]
Figured out why I was having issues testing locally. First was that I needed to switch the dashboard setting `Graph tooltip` to `Shared crosshair` (under dashboard settings -> `General` -> `Panel options` at bottom of page) (thanks to @leeoniya for this tip)

Then I needed to transform the time data (hard coded difference) due to what I believe to be time zone differences

**Comment by pbegg** [comment]
When is this likely to be available in grafana cloud?

**Pull request merged**